pkg/virt-launcher/virtwrap/agent-poller/agent_poller.go

```diff
@@ -31,265 +30,183 @@ import (
 )
 
-type AgentUpdateEvent struct {
-	InterfaceStatuses *[]api.InterfaceStatus
-	DomainName        string
-	OsInfo            *api.GuestOSInfo
-}
+// AgentCommand is a command exutable on guest agent
+type AgentCommand string
 
-type AgentPoller struct {
-	Connection      cli.Connection
-	VmiUID          types.UID
-	domainData      *DomainData
-	agentDone       chan struct{}
-	domainUpdate    chan *api.Domain
-	pollTime        time.Duration
-	agentUpdateChan chan AgentUpdateEvent
-}
+// Aliases for commands executed on guest agent
+// when updated to libvirt 5.6.0 this can change to libvirt types
+// Aliases are also used as keys to the store, it does not matter how the keys are named,
+// only whther it relates to the right data
+const (
+	GET_OSINFO     AgentCommand = "guest-get-osinfo"
+	GET_INTERFACES AgentCommand = "guest-network-get-interfaces"
+)
 
-type DomainData struct {
-	name       string
-	osInfo     api.GuestOSInfo
-	aliasByMac map[string]string
-	interfaces []api.InterfaceStatus
+// AgentUpdatedEvent fire up when data is changes in the store
+type AgentUpdatedEvent struct {
+	Type       AgentCommand
+	DomainInfo api.DomainGuestInfo
 }
 
-// Result for json unmarshalling
-type Result struct {
-	Interfaces []Interface `json:"return"`
+// AgentStore stores the agent data converted to api domain objects
+// it offers methods to get the data and fire up an event when there
+// is a change of the data
+type AgentStore struct {
+	store        sync.Map
+	AgentUpdated chan AgentUpdatedEvent
 }
 
-// Guest OS Info unmarshalling
-type ParsedGuestOsInfo struct {
-	Name          string `json:"name"`
-	KernelRelease string `json:"kernel-release"`
-	Version       string `json:"version"`
-	PrettyName    string `json:"pretty-name"`
-	VersionId     string `json:"version-id"`
-	KernelVersion string `json:"kernel-version"`
-	Machine       string `json:"machine"`
-	Id            string `json:"id"`
+// NewAgentAstore creates new agent store
+func NewAgentAstore() AgentStore {
+	return AgentStore{
+		store:        sync.Map{},
+		AgentUpdated: make(chan AgentUpdatedEvent, 10),
+	}
 }
 
-type GuestOSInfoResult struct {
-	OSInfo ParsedGuestOsInfo `json:"return"`
-}
+// Store save the value with a key to the storage, when there is a change in data
+// it fires up updated event
+func (s *AgentStore) Store(key AgentCommand, value interface{}) {
+	updated := false
 
-// Interface for json unmarshalling
-type Interface struct {
-	MAC  string `json:"hardware-address"`
-	IPs  []IP   `json:"ip-addresses"`
-	Name string `json:"name"`
-}
+	oldData, _ := s.store.Load(key)
+	if oldData == nil {
+		// nothing found, this means update
+		updated = true
+	} else if !reflect.DeepEqual(oldData, value) {
+		// of old data is not the same as new, this means update
+		updated = true
+	}
 
-// IP for json unmarshalling
-type IP struct {
-	IP     string `json:"ip-address"`
-	Type   string `json:"ip-address-type"`
-	Prefix int    `json:"prefix"`
-}
+	s.store.Store(key, value)
```

Comment:
Another thread could theoretically store the same key (and value) between calls to Load/Store here.  Which would lead to multiple events.  Probably not a big deal but a single mutex around both calls could protect from that.  Not currently an issue because only one poller thread.

Reply:
Hmm, wouldn't that actually be what we want? I mean, if multiple changes arise one after another, we wouldn't want to skip a change, or would we? Asked another way, would we want to prohibit changes occuring nearly at the same time, and if so, why?

Reply:
I think we may want to prevent getting two consecutive events for the same key+value.  And that is currently possible.

Reply:
Generating two consecutive call require two poll workers with the same configuration polling the same guest agent endpoint. Multiple pollers with the same guest agent endpoint is not the goal. There will be multiple workers, indeed. However each one will have different set of guest agent endpoints, and each will have different timeout.

There will be 4 workers, taken from PR: #2827 

* sys data: INTERFACES, OSINFO, TIMEZONE, HOSTNAME; queried every: 120s
* Agent: AGENT_VERSION; queried every: 300s
* Filesystem: FILESYSTEM; queried every: 300s
* User list: USERS: queried every: 10s

Enabling user to configure the choice of distribution does not make sense.
The calls are expensive and timeouts are chosen by trial in oVirt to not stress guest agent.

Furthermore timeouts can be tuned in PR: #2827 via parameter to virt-launcher, but I would suggest to not touch it.

The rule of thumb is, timeouts are selected to reflect how often the data changes on the host.